schemachanger: Implement `ALTER COLUMN ... SET NOT NULL`

Xiang-Gu opened the pull request:
This PR implements `ALTER TABLE t ALTER COLUMN c SET NOT NULL` in the declarative schema changer. 

The key change is to introduce the `ColumnNotNull` element for column NOT NULL constraint. Previously, this is recorded in a bool field in the `ColumnType` element. We will treat `ColumnNotNull` in a similar fashion as `CheckConstraint` where we indeed uses a dummy check constraint for validation.

The plan to address mixed version compatibility is laid out in comments in commit 2.

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95556)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/backfill.go:
```diff
@@ -1536,8 +1536,14 @@ func ValidateConstraint(
+					// Re-code the error to NotNullViolation if it's a NOT NULL constraint violation.
+					if err != nil && pgerror.GetPGCode(err) == pgcode.CheckViolation &&
+						ck.IsNotNullColumnConstraint() {
+						err = pgerror.WithCandidateCode(errors.UnwrapOnce(err), pgcode.NotNullViolation)
```
This `UnwrapOnce` is relying on the outermost error being the pgerror. That feels fragile. Maybe better to use `errors.If` to find the correct pgerror and replace it directly?

ajwerner commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -1752,7 +1752,10 @@ func (desc *Mutable) AddUniqueWithoutIndexMutation(
+	tbl catalog.TableDescriptor,
+	col catalog.Column,
+	validity descpb.ConstraintValidity,
+	constraintID descpb.ConstraintID,
```
comment on the semantics of the constraint ID here

ajwerner commented on pkg/sql/catalog/tabledesc/column.go:
```diff
@@ -33,8 +33,9 @@ var _ catalog.Column = (*column)(nil)
-	desc    *descpb.ColumnDescriptor
-	ordinal int
+	desc              *descpb.ColumnDescriptor
+	isBecomingNotNull bool
```
this is detritus, right?

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -277,11 +279,37 @@ type addColumnSpec struct {
-// addColumn is a helper function which adds column element targets and ensures
-// that the new column is backed by a primary index, which it returns.
+// addColumn adds a column as specified in the `spec`. It delegates most of the work
+// to addColumnIgnoringNotNull and handles NOT NULL constraint itself.
+// It is version aware in that it consults the current active cluster version and set all protobuf
+// fields for version-gating purposes.
 func addColumn(b BuildCtx, spec addColumnSpec, n tree.NodeFormatter) (backing *scpb.PrimaryIndex) {
```
You could say more here about what the version gate is all about, why it exists, how long it should last etc. If you stumbled upon this code and had not written it, would you know what it meant?

ajwerner commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -2676,7 +2676,7 @@ INSERT INTO t_add_column_not_null VALUES (1), (2), (3)
 ALTER TABLE t_add_column_not_null ADD COLUMN j INT NOT NULL DEFAULT (NULL:::INT)
 
-statement error pgcode 23502 null value in column "j" violates not-null constraint
+statement error pgcode 23502 .*
```
I think you can do better than .*. Even if there are two different possibilities, it's a regexp, you can use `|`

ajwerner reviewed (commented):
Nice, the TestXiang stuff, of course, should use some cleanup. 

Can you add some end-to-end tests?

Xiang-Gu committed:
schemachanger: Turn `SET NOT NULL` on by default and add tests

Xiang-Gu committed:
schemachanger: Introduce `ColumnNotNull` element

This commit introduce a new element protobuf `ColumnNotNull`, which
will be used in later commit to re-implement how we handle NOT NULL
constraint of a column.

Xiang-Gu committed:
schemachanger: Re-implement logic handling NOT NULL constraint

This commit uses the newly introduced element `ColumnNotNull` and
re-implement logic handling NOT NULL constraint of a column in the
declarative schema changer.

The key idea is to treat NOT NULL as a functionally equivalent
CHECK constraint, (col_name IS NOT NULL), and use existing validation
logic for CHECK to validate the NOT NULL constraint.

This commit is a implementation change for something we already support
in the declarative schema changer, so the way we achieve compatibility
in mixed version state is to properly version gate the builder and
mutation execution. The detailed plan can be found in the comment inside
`ColumnType` element.

Xiang-Gu commented:
@ajwerner I've addressed your early feedback. I'm investigating the one and last test failure in CI. At the same time, can you give it another look? Thank you a lot

Xiang-Gu commented:
This is officially ready for review!

ajwerner reviewed (commented):
Very nice!

> Reviewed 7 of 32 files at r1, 8 of 27 files at r2, 50 of 168 files at r3, 7 of 15 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-:-NM_-0HXCgbfoKx9UP_B:b-j7qk9t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 449 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMZwY0H9ACM0qfoau4j:-NMZwY0H9ACM0qfoau4k:bvy1g2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scdecomp/decomp.go#L449)):*
> ```Go
> 			//ConstraintID: tbl.GetNextConstraintID(),
> 		})
> 		//tbl.TableDesc().NextConstraintID++
> ```

detritus?



_Code quote:_
```Go
			//ConstraintID: tbl.GetNextConstraintID(),
		})
		//tbl.TableDesc().NextConstraintID++
	}
```
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/helpers.go` line 222 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMZwopp6rkutNsCLXm3:-NMZwopp6rkutNsCLXm4:b-1y1rmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scexec/scmutationexec/helpers.go#L222)):*
> ```Go
> }
> 
> func enqueueAddNotNullMutation(
> ```

nit: this subtraction is getting silly, and nothing seems to return an error. We can use generics now. What if you do:

```
func addMutation[M any](f func(M, descpb.DescriptorMutation_Direction), m M, dir descpb.DescriptoMutation_Direction) {
    f(tbl, m, dir)
    tbl.NextMutationID--
}
```


Then you can call this with:
```
addMutation(tbl.AddNotNullMutation, ck, descpb.DescriptorMutation_ADD)
```
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_set_not_null` line 7 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NM_-At8144RVdtgssIo:-NM_-At8144RVdtgssIp:b-sid2df) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_set_not_null#L7)):*
> ```
> +object {100 101 t} -> 104
> 
> test
> ```

can you add some statements during this which attempt to write null values and fail at the appropriate stages? 


<!-- Sent from Reviewable.io -->

postamar commented:
Can you hold off before merging this please? I would like to get https://github.com/cockroachdb/cockroach/pull/95631 in first, because otherwise a lot of rebasing is needed on my end and time is running out before I go on leave. Thanks!

Xiang-Gu committed:
scmutationexec: Use Go generics to replace a bunch of repetitive funcs

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/release_22_2/testdata/deprules:
```diff
@@ -2023,7 +2120,7 @@ deprules
   query:
     - $descriptor[Type] IN ['*scpb.Database', '*scpb.Schema', '*scpb.View', '*scpb.Sequence', '*scpb.Table', '*scpb.EnumType', '*scpb.AliasType', '*scpb.CompositeType']
-    - $idx-or-col[Type] IN ['*scpb.Column', '*scpb.PrimaryIndex', '*scpb.SecondaryIndex', '*scpb.TemporaryIndex', '*scpb.UniqueWithoutIndexConstraint', '*scpb.CheckConstraint', '*scpb.ForeignKeyConstraint']
+    - $idx-or-col[Type] IN ['*scpb.Column', '*scpb.PrimaryIndex', '*scpb.SecondaryIndex', '*scpb.TemporaryIndex', '*scpb.UniqueWithoutIndexConstraint', '*scpb.CheckConstraint', '*scpb.ForeignKeyConstraint', '*scpb.ColumnNotNull']
```
None of the test output in the release_22_2 package should change, ever. Also, I don't think that foreign key constraints are featured in 22.2 so I filed an issue: https://github.com/cockroachdb/cockroach/issues/95849

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_not_null.go:
```diff
@@ -0,0 +1,51 @@
+			))
+		}
+	})
+}
```
nit: Should this be in `alter_table_add_constraint.go` instead? I realize there's no actual ADD CONSTRAINT syntax for NOT NULL, so you be the judge.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_not_null.go:
```diff
@@ -0,0 +1,51 @@
+// Copyright 2022 The Cockroach Authors.
```
s/2022/2023

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/current/op_drop.go:
```diff
@@ -172,6 +172,7 @@ func init() {
 				(*scpb.CheckConstraint)(nil),
 				(*scpb.ForeignKeyConstraint)(nil),
 				(*scpb.UniqueWithoutIndexConstraint)(nil),
+				(*scpb.ColumnNotNull)(nil),
```
Perhaps we should have a filter in the helpers for non-index-backed constraints?

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go:
```diff
@@ -97,4 +97,33 @@ func init() {
+	// "too late" (i.e. when column transitions to ABSENT) is that the
+	// constraint would still be visible and enforced while the column is not writable.
+	// This can cause concurrent WRITE to see and attempt to uphold the constraint,
+	// but it wouldn't be able to find the column.
```
Thanks for the explanation, I get why it needs to be in the same stage. Any reason why it's the not-null before the column and not the other way round? Not that it matters, just curious.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -182,10 +182,36 @@ message ColumnType {
+  //  - Delete `is_23_1_cluster_version_active` field from protobuf.
+  // This plan will guarantee compatibility in all possible mixed version states
+  // as well as restoring from a old backup.
+  bool deprecated_is_nullable = 5;
```
You can deprecate fields in protobuf with `[deprecated=true]`. Also add `// Deprecated.` in the comment block above for a pretty strike-through visual effect in GoLand.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -182,10 +182,36 @@ message ColumnType {
+  // V24_1:
+  //  - Delete `is_23_1_cluster_version_active` field from protobuf.
+  // This plan will guarantee compatibility in all possible mixed version states
+  // as well as restoring from a old backup.
```
Comments such as these don't typically age well. I recommend you use language which is more non-committal and less subject to current assumptions as to what the future will look like. You're far better off filing tickets in github to track future work.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -182,10 +182,36 @@ message ColumnType {
+  // This plan will guarantee compatibility in all possible mixed version states
+  // as well as restoring from a old backup.
+  bool deprecated_is_nullable = 5;
+  bool is_23_1_cluster_version_active = 11;
```
Consider renaming this to `is_element_created_in_23_1_or_later`? This field is a neat solution to the problem at hand, which I'd find mildly distasteful if it weren't for the fact that we need to migrate to a new model for ColumnType in this release anyway as per https://github.com/cockroachdb/cockroach/issues/92605

postamar replied:
An alternative to consider, which might age better: define a
```
message ElementCreationMetadata {
  bool in_23_1_or_later = 1;
}
```
and embed it into `ColumnType`. Then in scdecomp provide a helper which generates a `scpb.ElementCreationMetadata` instance based on the cluster settings. This way this logic can be re-used and adapted for later versions and for other elements as well.

Another alternative: have this metadata in the `TargetMetadata` message instead.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -181,10 +182,36 @@ message ColumnType {
+  //  - Start deprecating `is_nullable` (rename it to `deprecated_is_nullable`);
+  //  - Introduce `is_23_1_cluster_version_active`;
+  //  - Add a version gate in the builder for `ADD COLUMN` (see addColumn function)
+  //    to set `is_23_1_cluster_version_active` if cluster version V23_1 is active.
```
Shouldn't `scdecomp` be doing that as well?

postamar reviewed (commented):
I did a pass, this approach seems sound.

Xiang-Gu reviewed (commented):
@ajwerner @postamar Thank you both for the feedback. I've addressed them all. I will hold off merging until #95631 is in, but in the meantime, I'd like to hear any other comments/feedback you still have for this PR. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-:-NMfS3vo0WcaGJJCNYjf:b-8dajtf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_not_null.go` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMGRuqaDeeT0-jUgdkk-r4-1:-NMeiNtEE30F2V07d43l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_not_null.go#L1)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

s/2022/2023
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_not_null.go` line 51 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMGRuqaDeeT0-jUgdkk-r4-51:-NMeiOuo6W9WwH2VfTkL:b-blrblp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_set_not_null.go#L51)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: Should this be in `alter_table_add_constraint.go` instead? I realize there's no actual ADD CONSTRAINT syntax for NOT NULL, so you be the judge.
</blockquote></details>

I'd like to keep as is bc 1. it's less boilerplate code added (all the nested min supported version control for add constraints), and 2. we don't have `SET NOT NULL NOT VALID`as other constraints.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 449 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMZwY0H9ACM0qfoau4j:-NM_uZFs8JE_8sKpnk1k:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scdecomp/decomp.go#L449)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

detritus?
</blockquote></details>

done



_Code quote:_
```Go
			ColumnID: col.GetID(),
			//ConstraintID: tbl.GetNextConstraintID(),
```
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/helpers.go` line 222 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMZwopp6rkutNsCLXm3:-NMdmjs6E7na8k1fLTet:bix86o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scexec/scmutationexec/helpers.go#L222)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: this subtraction is getting silly, and nothing seems to return an error. We can use generics now. What if you do:

```
func addMutation[M any](f func(M, descpb.DescriptorMutation_Direction), m M, dir descpb.DescriptoMutation_Direction) {
    f(tbl, m, dir)
    tbl.NextMutationID--
}
```


Then you can call this with:
```
addMutation(tbl.AddNotNullMutation, ck, descpb.DescriptorMutation_ADD)
```

</blockquote></details>

This is a good idea! Done
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 209 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMBA12LJbgA9llXmquK-r3-209:-NMfSYn61-UuPronf6YS:bfipmz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/61e32b741a5b41c41958bda3fa0f2f3534751594/pkg/sql/schemachanger/scpb/elements.proto#L209)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Comments such as these don't typically age well. I recommend you use language which is more non-committal and less subject to current assumptions as to what the future will look like. You're far better off filing tickets in github to track future work.
</blockquote></details>

done, filed https://github.com/cockroachdb/cockroach/issues/95893
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 210 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMBA12LJbgA9llXmquK-r3-210:-NMfU8lw0v8Kl0tj33ZN:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/61e32b741a5b41c41958bda3fa0f2f3534751594/pkg/sql/schemachanger/scpb/elements.proto#L210)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You can deprecate fields in protobuf with `[deprecated=true]`. Also add `// Deprecated.` in the comment block above for a pretty strike-through visual effect in GoLand.
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 211 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMBA12LJbgA9llXmquK-r3-211:-NMfMLWOAPMYBIGBXJkn:b-e21eih) ([raw file](https://github.com/cockroachdb/cockroach/blob/61e32b741a5b41c41958bda3fa0f2f3534751594/pkg/sql/schemachanger/scpb/elements.proto#L211)):*
> message ElementCreationMetadata {
>   bool in_23_1_or_later = 1;
> }
I like this idea. I did some plumbing so that we have the cluster version in `scdecomp`. 
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 196 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMBA12LJbgA9llXmquK-r4-196:-NMetRrKC0svOzReje5L:b-wla3i7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/scpb/elements.proto#L196)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't `scdecomp` be doing that as well?
</blockquote></details>

it's okay bc we have a minSupportedVersion on elements and any elements not supported in current cluster version will be filtered out. In fact, I think this makes it easier to think about: In the builder, we use the version gate to only set this newly added field `is_23_1_cluster_version_active`. We still decomp and add/drop newly introduced elements (in addition to what we've already been doing) while rest assured that it will be filtered out anyway if they're not supported.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMGBeBlyv1k6mnYteA2-r3-115:-NMfN99x4kMQr9ihUvBj:b3mv6o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/61e32b741a5b41c41958bda3fa0f2f3534751594/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L115)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Thanks for the explanation, I get why it needs to be in the same stage. Any reason why it's the not-null before the column and not the other way round? Not that it matters, just curious.
</blockquote></details>

I think that'd be fine. All we need here is same-stage.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/op_drop.go` line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMGBeBlyv1k6mnYteA3-r3-175:-NMfNioGBAEC_RB1r8bF:b2geonx) ([raw file](https://github.com/cockroachdb/cockroach/blob/61e32b741a5b41c41958bda3fa0f2f3534751594/pkg/sql/schemachanger/scplan/internal/rules/current/op_drop.go#L175)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Perhaps we should have a filter in the helpers for non-index-backed constraints?
</blockquote></details>

We do! `IsSupportedNonIndexBackedConstraint`



_Code quote:_
```Go
			constraint.Type(
				(*scpb.CheckConstraint)(nil),
				(*scpb.ForeignKeyConstraint)(nil),
				(*scpb.UniqueWithoutIndexConstraint)(nil),
				(*scpb.ColumnNotNull)(nil),
```
___
*[`pkg/sql/schemachanger/scplan/internal/rules/release_22_2/testdata/deprules` line 2123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMVUqmR49HWD0ngCYbJ-r2-2123:-NMfNsUZ0N1LveMhYlsz:b-c0z4ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f0f3ef7bc2795c216bca119c0c0eb8232349164/pkg/sql/schemachanger/scplan/internal/rules/release_22_2/testdata/deprules#L2123)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

None of the test output in the release_22_2 package should change, ever. Also, I don't think that foreign key constraints are featured in 22.2 so I filed an issue: https://github.com/cockroachdb/cockroach/issues/95849
</blockquote></details>

yeah, we are aware of this and thanks for filing the issue.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_set_not_null` line 7 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NM_-At8144RVdtgssIo:-NMfNzB6EAjFfOWXeSES:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/5444e27379f29098f47076f8636a8abe2fd7564e/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_set_not_null#L7)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you add some statements during this which attempt to write null values and fail at the appropriate stages? 
</blockquote></details>

Done!
___
*[`pkg/sql/catalog/tabledesc/column.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95556#-NMBA12LJbgA9llXmqu0-r1-37:-NMdmvNQ6FaAKCSzulJr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9213c205ed565b97b93210dce3bdb3241db427a0/pkg/sql/catalog/tabledesc/column.go#L37)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is detritus, right?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/catalog/tabledesc/validate.go:
```diff
@@ -898,6 +898,11 @@ func (desc *wrapper) validateConstraintNamesAndIDs(vea catalog.ValidationErrorAc
 	for _, c := range constraints {
+		if c.AsCheck() != nil && c.AsCheck().IsNotNullColumnConstraint() {
+			// Relax validation for NOT NULL constraint when disguised as CHECK
+			// constraint, because they don't a constraintID.
```
nit: they don't *have* a constraintID

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -278,11 +281,33 @@ type addColumnSpec struct {
 
-// addColumn is a helper function which adds column element targets and ensures
-// that the new column is backed by a primary index, which it returns.
+// addColumn adds a column as specified in the `spec`. It delegates most of the work
+// to addColumnIgnoringNotNull and handles NOT NULL constraint itself.
+// It contains version gates to help ensure compatibility in mixed version state.
+// Read comments in `ColumnType` message in `elements.proto` for details.
 func addColumn(b BuildCtx, spec addColumnSpec, n tree.NodeFormatter) (backing *scpb.PrimaryIndex) {
+	backing = addColumnIgnoringNotNull(b, spec, n)
+	if spec.notNull {
+		cnne := scpb.ColumnNotNull{
+			TableID:  spec.tbl.TableID,
+			ColumnID: spec.col.ColumnID,
+		}
+		if backing != nil {
+			cnne.IndexIDForValidation = backing.IndexID
+		}
+		b.Add(&cnne)
+	}
+	return backing
+}
+
+// addColumnIgnoringNotNull is a helper function which adds column element targets and ensures
+// that the new column is backed by a primary index, which it returns.
+func addColumnIgnoringNotNull(
```
Does anybody call `addColumnIgnoringNotNull` besides `addColumn`? I didn't see any callers. Given that, it feels like this should just be logic in `addColumn`.

Xiang-Gu replied:
done!

ajwerner reviewed (approved):
LGTM

Xiang-Gu commented:
TFTR!

bors r+
